fastmri_recon/data/datasets/multicoil/preprocessing.py

```diff
@@ -33,6 +36,60 @@ def from_kspace_to_masked_kspace_and_mask(images, kspaces):
             return (kspaces_channeled, mask, smaps), images_channeled
     return from_kspace_to_masked_kspace_and_mask
 
+
+def non_cartesian_from_kspace_to_nc_kspace_and_traj(
+        nfft_ob,
+        image_size,
+        acq_type='radial',
+        scale_factor=1,
+        compute_dcomp=False,
+        parallel=False,
+        **acq_kwargs
+):
+    def from_kspace_to_nc_kspace_and_traj(images, kspaces):
+        if acq_type == 'radial':
+            traj = get_radial_trajectory(image_size, **acq_kwargs)
+        elif acq_type == 'cartesian':
+            traj = get_debugging_cartesian_trajectory()
+        else:
+            raise NotImplementedError(f'{acq_type} dataset not implemented yet.')
+        interpob = nfft_ob._extract_nufft_interpob()
+        nufftob_forw = kbnufft_forward(interpob)
+        nufftob_back = kbnufft_adjoint(interpob)
+        dcomp = calculate_radial_dcomp_tf(
+            interpob,
+            nufftob_forw,
+            nufftob_back,
+            traj[0],
+        )
+        traj = tf.repeat(traj, tf.shape(images)[0], axis=0)
+        orig_image_channels = tf_unmasked_adj_op_non_cartesian(kspaces[..., None])[..., 0]
+        nc_kspace = nufft(nfft_ob, orig_image_channels, traj, image_size)
+        nc_kspace_scaled = nc_kspace * scale_factor
+        images_scaled = images * scale_factor
+        images_channeled = images_scaled[..., None]
+        nc_kspaces_channeled = nc_kspace_scaled[..., None]
+        orig_shape = tf.ones([tf.shape(kspaces)[0]], dtype=tf.int32) * tf.shape(kspaces)[-1]
+        extra_args = (orig_shape,)
+        dcomp = tf.ones([tf.shape(kspaces)[0], tf.shape(dcomp)[0]], dtype=dcomp.dtype) * dcomp[None, :]
+        if compute_dcomp:
+            extra_args += (dcomp,)
+        if parallel:
```

Comment:
let's not have parallel for non cartesian, this will never be used I think...

Reply:
To be honest, parallel is same as Calibrationless Reconstruction. This is known to preserve phase and thus more helpful from what I remember

Reply:
hmmm I see but then we need to have coil selection implemented here, which is something I want to avoid for now. Let's cross that bridge when we get there (it works in cartesian case so when we need we could use it).